fix: Fixed a few memory leak cases when shutting down NetworkManager during Incoming Message Queue processing

ShadauxCat opened the pull request:
Fixed a few memory leak cases when shutting down NetworkManager during Incoming Message Queue processing.

MTT-1440

(I'm not sure if I should update changelog.md since it still says [Unreleased] and I'm not sure if I should put this in that section or make a new one for 1.0.1)

## Changelog

### com.unity.netcode.gameobjects
- Fixed: Fixed a few memory leak cases when shutting down NetworkManager during Incoming Message Queue processing.

## Testing and Documentation

* Includes integration tests.
* No documentation changes or additions were necessary.

ShadauxCat committed:
fix: Fixed memory leak on shutdown

ShadauxCat committed:
fix: Fixed a few memory leak cases when shutting down NetworkManager during Incoming Message Queue processing

ShadauxCat committed:
standards check

ShadauxCat committed:
Fixed test not cleaning up quite correctly.

mattwalsh-unity commented on com.unity.netcode.gameobjects/Runtime/Messaging/MessagingSystem.cs:
```diff
@@ -113,6 +114,16 @@ public void Dispose()
                 CleanupDisconnectedClient(kvp.Key);
             }
+
+            // If there's a message currently being processed, it'll dispose itself, so we can't dispose here or that'll
```
Given we're single-threaded, is it possible we could be in the middle of processing a message and get here too?

ShadauxCat replied:
So just to put this here even though we discussed it on slack... It's not an issue of concurrency, it's an issue of call stack.

Three messages come in: FooMessage, ClientRpc, BarMessage

We process FooMessage and FooMessage is Dispose()d

We start to process ClientRpc

    ClientRpc message calls NetworkManager.Shutdown(), which calls MessagingSystem.Dispose()
    
After returning from ClientRpc, the ClientRpc message is Dispose()d.

Since MessagingSystem has now been Dispose()d as well, we can no longer process BarMessage, so we exit the loop.

Thus:
- We can't Dispose() FooMessage because it was already Dispose()d
- We can't Dispose() the current message being processed (ClientRpc) because HandleMessage is above us in the call stack and will Dispose() it before it returns
- We must dispose BarMessage because HandleMessage will not get to that one as we're Dispose()ing the list that contains it.

That's what this comment means - it's not a threading concurrency, it's just that Shutdown() can be called from within HandleMessage(), and if that happens, the message currently being processed can't be Dispose()d.

andrews-unity committed:
Merge branch 'develop' into fix/fix_memory_leak_on_shutdown

andrews-unity commented on com.unity.netcode.gameobjects/Runtime/Messaging/MessagingSystem.cs:
```diff
@@ -461,16 +476,16 @@ internal unsafe void ProcessSendQueues()
                     {
                         m_MessageSender.Send(clientId, queueItem.NetworkDelivery, queueItem.Writer);
+
+                        for (var hookIdx = 0; hookIdx < m_Hooks.Count; ++hookIdx)
```
I assume this is not actually required for this PR? just fixed along the way?

ShadauxCat replied:
I honestly don't remember... this PR is over a month old now. I'm pretty sure I ran into an issue with the test case that exposed this as a bug and had to fix it for the test case to pass, but it's been too long for me to be confident in that.

andrews-unity committed:
Merge branch 'develop' into fix/fix_memory_leak_on_shutdown

andrews-unity commented:
So I have a question and this just for my understanding... Is there any reason why we just have some sort of flag that says currently we are processing messages and if you call dispose while we are processing these messages or Shutdown on `NetworkManager`, we will set `m_Dispose `to true and then once we are done processing all the messages we will dispose the `NetworkManager` ? Because I think in some cases you may want to process all the pending messages before you clean up stuff and then we don't have this magic index etc....

mattwalsh-unity committed:
Merge branch 'develop' into fix/fix_memory_leak_on_shutdown

github-actions[bot] committed:
Merge develop into fix/fix_memory_leak_on_shutdown

unity-cla-assistant commented:
[![CLA assistant check](https://cla-assistant.cds.internal.unity3d.com/pull/badge/not_signed)](https://cla-assistant.cds.internal.unity3d.com/Unity-Technologies/com.unity.netcode.gameobjects?pullRequest=1323) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.cds.internal.unity3d.com/Unity-Technologies/com.unity.netcode.gameobjects?pullRequest=1323) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.cds.internal.unity3d.com/check/Unity-Technologies/com.unity.netcode.gameobjects?pullRequest=1323) it.</sub>

andrews-unity committed:
Merge branch 'develop' into fix/fix_memory_leak_on_shutdown

andrews-unity committed:
Merge branch 'develop' into fix/fix_memory_leak_on_shutdown

andrews-unity committed:
Merge branch 'develop' into fix/fix_memory_leak_on_shutdown

andrews-unity committed:
Merge branch 'develop' into fix/fix_memory_leak_on_shutdown

ShadauxCat committed:
-Reworked the way shutdown is handled so it can't be processed while handling incoming messages.
-Added tests for both ways to shut down to make sure they both work correctly
-Fixed an issue in the snapshot system that was causing the new tests to fail

ShadauxCat commented on com.unity.netcode.gameobjects/Runtime/Core/SnapshotSystem.cs:
```diff
@@ -491,9 +491,20 @@ internal void SpawnObject(SnapshotSpawnCommand spawnCommand, ulong srcClientId)
-                var networkObject = m_NetworkManager.SpawnManager.CreateLocalNetworkObject(false,
-                    spawnCommand.GlobalObjectIdHash, spawnCommand.OwnerClientId, (spawnCommand.ParentNetworkId == spawnCommand.NetworkObjectId) ? spawnCommand.NetworkObjectId : spawnCommand.ParentNetworkId, spawnCommand.ObjectPosition,
-                    spawnCommand.ObjectRotation);
+                NetworkObject networkObject;
```
This fix is technically not a part of this change, but the added tests won't pass without this change because it causes warnings to get logged.

andrews-unity committed:
Merge branch 'develop' into fix/fix_memory_leak_on_shutdown

ShadauxCat committed:
standards

ShadauxCat committed:
Fixed broken test, slight naming change to be more accurate.

ashwinimurt commented:
I will be updating the changelog.
